### PR TITLE
Feature/project dependency precedence (without disableProjectPriority option)

### PR DESCRIPTION
**Pull request by idlsoft**
Thank you very much for your contribution. We highly appreciate the time
and effort you put into pull request. Before you submit the pull request,
please review the [Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#contribution-workflow).

Make sure you check the following boxes.

For all changes:

- [x] Sign the [Gradle CLA](http://gradle.org/contributor-license-agreement/).

For all non-trivial changes that modify the behavior or public API:

- [x] Before submitting a pull request, you started a discussion on the
[Gradle developer list](https://groups.google.com/forum/#!forum/gradle-dev)
or can reference a [JIRA issue](https://issues.gradle.org/secure/Dashboard.jspa).
- [x] You considered writing a design document. A design document can be
brief but explains the use case or problem you are trying to solve,
touches on the planned implementation approach as well as the test cases
that verify the behavior. Optimally, design documents should be submitted
as a separate pull request. You can find [samples](https://github.com/gradle/gradle/tree/master/design-docs)
in the Gradle GitHub repository. Please let us know if you need help with
creating the design document. We are happy to help!
- [x] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior. Before submitting the pull request
you ran a build on your local machine via the command `./gradlew quickCheck`
and the tests for the modified module.
- [x] The pull request updates the Gradle documentation like user guide,
DSL reference and Javadocs where applicable.

### CODE REVIEW

**Comment by idlsoft** [comment]
This is an alternative implementation to https://github.com/gradle/gradle/pull/564

**Comment by bigdaz** [comment]
I recently made changes to `DependencySubstitutionRulesIntegrationTest` so that it uses the `ResolveTestFixture` to validate the resolved graph. Your merge commit removed a lot of this work. Instead, the new tests should be written to be consistent with the refactored version, using `ResolveTestFixture`.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyResolveRulesIntegrationTest.groovy [hunk 1]
```diff
@@ -66,6 +67,83 @@ class DependencyResolveRulesIntegrationTest extends AbstractIntegrationSpec {
...
+            mavenRepo.module("org.utils", "api", forcedVersion).publish()
+        }
+        mavenRepo.module("org.utils", "api-ext", '1.0').dependsOn("org.utils", "api", versionExternal).publish()
+        settingsFile << 'include "api", "moduleA", "moduleB"'
+
+        String apiDependency = forcedVersion==null && forcedVersion!='project'?
+            'conf project(":api")' : "conf ('org.utils:api:$forcedVersion') { force = true }"
+
+        String substitutionSetting = forcedVersion == 'project'?
+            "dependencySubstitution { substitute module('org.utils:api') with project(':api') }" : ''
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyResolveRulesIntegrationTest.groovy]: This is never going to be used. We should have a test that demonstrates using dependency substitution to choose an external dependency over a project dependency.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencySubstitutionRulesIntegrationTest.groovy [hunk 2]
```diff
@@ -453,6 +453,18 @@ class DependencySubstitutionRulesIntegrationTest extends AbstractIntegrationSpec
                 configurations.conf.resolutionStrategy.dependencySubstitution {
                     substitute project(":api") with module("org.utils:api:1.5")
                 }
+
+                task check(dependsOn: configurations.conf) << {
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencySubstitutionRulesIntegrationTest.groovy]: Should be using `ResolveTestFixture` here.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VersionConflictResolutionIntegrationTest.groovy [hunk 3]
```diff
@@ -636,6 +636,12 @@ repositories {
 dependencies {
     compile "org:other:1.7"
 }
+
+configurations.all {
+    resolutionStrategy.dependencySubstitution {
+        substitute project('test') with module('org:test:2.1')
+    }
+}
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VersionConflictResolutionIntegrationTest.groovy]: This effectively breaks the use case described in GRADLE-2920, where the user wants to find the latest published version of the current module. To make this work, the test needs to demonstrate that this version can be _discovered_, not just resolved.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ProjectDependencyForcingResolver.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public <T extends ComponentResolutionState> T select(Collection<? extends T> candidates) {
+        // the collection will only be initialized if more than one project candidate is found
+        Collection<T> projectCandidates = null;
+        T foundProjectCandidate = null;
+        for(T candidate: candidates) {
+            ComponentResolveMetaData metaData = candidate.getMetaData();
+            if (metaData!=null && metaData.getComponentId() instanceof ProjectComponentIdentifier) {
+                ModuleVersionIdentifier moduleId = metaData.getId();
+                if (callingModule!=null && (callingModule.getGroup().equals(moduleId.getGroup())
+                && callingModule.getName().equals(moduleId.getName()))) continue;
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ProjectDependencyForcingResolver.java]: This should be refactored for clarity: it's very hard to see what's going on at a glance.
(Also this _should_ fail checkstyle, if it doesn't)

Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ProjectDependencyForcingResolver.java]: Also, should be able to use `moduleId.getModule().equals(callingModule)`

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ProjectDependencyForcingResolver.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                        projectCandidates = new ArrayList<T>();
+                        projectCandidates.add(foundProjectCandidate);
+                    }
+                    projectCandidates.add(candidate);
+                }
+            }
+        }
+        if (projectCandidates!=null) {
+            return delegate.select(projectCandidates);
+        }
```
Comment (bigdaz) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/ProjectDependencyForcingResolver.java]: What _exactly_ is the logic in this class doing? I'm not clear why we are excluding the current project from the candidates. Some clarifying comments and refactoring might help.

File: subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyResolveRulesIntegrationTest.groovy [hunk 6]
```diff
@@ -66,6 +67,83 @@ class DependencyResolveRulesIntegrationTest extends AbstractIntegrationSpec {
...
+    {
+        mavenRepo.module("org.utils", "api", versionExternal).publish()
+        if (forcedVersion!=null && forcedVersion!='project' && forcedVersion!=versionExternal) {
+            mavenRepo.module("org.utils", "api", forcedVersion).publish()
+        }
+        mavenRepo.module("org.utils", "api-ext", '1.0').dependsOn("org.utils", "api", versionExternal).publish()
+        settingsFile << 'include "api", "moduleA", "moduleB"'
+
+        String apiDependency = forcedVersion==null && forcedVersion!='project'?
+            'conf project(":api")' : "conf ('org.utils:api:$forcedVersion') { force = true }"
```
Comment (bigdaz) [inline subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyResolveRulesIntegrationTest.groovy]: _Direct_ dependencies with `force=true` have a special [Conflict resolver](https://github.com/gradle/gradle/blob/f20fb44a5898b47ec11dc8dd4af5f685fd7350c4/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/DependencyGraphBuilder.java#L893-L893). I suspect that this is only working for direct dependencies, and not for versions forced via dependency resolve rules (or replaced via dependency substitution rules).

I think that the processing is going to need to be more sophisticated to make this "just work", and we should clearly define all of the cases that should be handled: direct dependencies with 'force', published dependencies that have 'force', versions forced with dependency resolve rules, versions substituted with dependency substitution rules, etc, etc.

**Comment by bigdaz** [comment]
Sorry not to have been more involved in the conversation earlier. I'm a bit concerned about the breaking nature of this change, even if it is going to match user expectations for most builds. I don't think this PR has adequate coverage for all of the ways that the user can set expectations of the resolved version.

In order to make this work, we will need to consider the set of candidates in conflict resolution, and consider the `ComponentSelectionReason` of each. We'll need to come up with a set of rules to apply that make sense for a variety of candidates, which can be a mix of project components and external components, each with different forced/requested/selected-by-rule/... selection reasons.

It might be easier to go back to adding the 'prefer project components over exernal components' as an explicit conflict resolution option (`ResolutionStrategy.preferProjectComponents()`), and then refine this behaviour over time until we are ready to switch to it as the default.

**Comment by idlsoft** [comment]
This PR was indeed a bit rushed, I didn't realize how much the unit-tests have changed since the original submission. I'll try to find some time this week to merge this properly.

The PRchange is the introduction of `ProjectDependencyForcingResolver`, which is chained before  `LatestModuleConflictResolver`. It picks out project dependencies from the list of conflicts, and tries to resolve among them. In the original PR that was all it did. In this variant it excludes the calling module in order to accommodate scenarios that would otherwise use `disableProjectPriority`. 

My personal preference is to be explicit, but concerns about adding more flags to `ResolutionStrategy` were raised, so this was an attempt to avoid it.

If there is consensus that an explicit `disableProjectPriority` is preferable, we should probably close this PR and continue the review of the original.

**Comment by bigdaz** [comment]
Let's go with an explicit _opt-in_ for the new behaviour: I recommend
`ResolutionStrategy.preferProjectModules`. We should then map this to a new
`ConflictResolution` subtype, and add the
`ProjectDependencyForcingResolver` when it is set. This will have no chance
of breaking existing builds, and will allow us to separately discuss the
way we can move toward making this the default behaviour.

There are quite a few mechanisms that builds can use to explicitly specify
the dependency version, including `ResolutionStrategy.force`,
`ResolutionStrategy.eachDependency`,
`ResolutionStrategy.dependencySubstitution`, `Dependency.force` and the
'force' attribute in Ivy.xml. I think we need to consider how these will be
treated with respect to conflict resolution. This should be done in a
design spec, and each will require explicit test coverage before I'd be
happy enabling this behaviour by default.

On Tue, 3 May 2016 at 08:30 Pepper Lebeck-Jobe <notifications@github.com>
wrote:

> Assigned #631 <https://github.com/gradle/gradle/pull/631> to @bigdaz
> <https://github.com/bigdaz>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/gradle/gradle/pull/631#event-648916526>
>
-- 


Daz DeBoer

Principal Software Engineer

W. www.gradle.com

**Comment by idlsoft** [comment]
I'll try to create a new PR this week.

**Comment by idlsoft** [comment]
@bigdaz I'm not sure about adding a new `ConflictResolution` subtype.
It makes one particular scenario tricky:
```groovy
resolutionStrategy {
     failOnVersionConflict()
     preferProjectModules(true)
}
```
Should it just override `StrictConflictResolution`? Probably not. Should it check the type it's overriding? Also not pretty. 
In the original PR I added a boolean to LatestConflictResolution, and would only set it if that was the current `ConflictResolution`. I makes prioritizing `StrictConflictResolution` more natural.

**Pull request closed without merging**